app/src/main/java/org/metabrainz/mobile/activity/SearchActivity.java

```diff
@@ -0,0 +1,290 @@
+
+        chooseAdapter();
+        recyclerView.setAdapter(adapter);
+        recyclerView.setLayoutManager(new LinearLayoutManager(this));
+        viewModel = ViewModelProviders.of(this).get(SearchViewModel.class);
+
+        viewModel.prepareSearch(query, entity);
+        doSearch(query);
+
+        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
+        noRes = findViewById(R.id.noresults);
+    }
+
+    @Override
+    protected void onNewIntent(Intent intent) {
+        setIntent(intent);
+    }
+
+    /*private void displayArtistResultsView() {
+
+        artistResultsView.setAdapter(new ArtistSearchAdapter(SearchActivity.this, artistSearchResults));
+        //artistResultsView.setOnItemClickListener(new ArtistItemClickListener());
+        artistResultsView.setVisibility(View.VISIBLE);
+
+        if (artistSearchResults.isEmpty())
+            noRes.setVisibility(View.VISIBLE);
+        else {
+            noRes.setVisibility(View.GONE);
+            //saveQueryAsSuggestion();
+        }
+    }
+
+    private void displayRGResultsView() {
+        ListView rgResultsView = (ListView) findViewById(R.id.searchres_rg_list);
+        rgResultsView.setAdapter(new RGSearchAdapter(SearchActivity.this, rgSearchResults));
+        //rgResultsView.setOnItemClickListener(new RGItemClickListener());
+        rgResultsView.setVisibility(View.VISIBLE);
+
+        if (rgSearchResults.isEmpty()) {
+            noRes.setVisibility(View.VISIBLE);
+        } else {
+            noRes.setVisibility(View.GONE);
+            //saveQueryAsSuggestion();
+        }
+    }*/
+
+    /* private void saveQueryAsSuggestion() {
+        if (App.getUser().isSearchSuggestionsEnabled()) {
+            SearchRecentSuggestions suggestions = new SearchRecentSuggestions(this, SuggestionProvider.AUTHORITY,
+                    SuggestionProvider.MODE);
+            suggestions.saveRecentQuery(searchTerm, null);
+        }
+    }*/
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        getMenuInflater().inflate(R.menu.search, menu);
+        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
+        searchView = (SearchView) menu.findItem(R.id.search_view).getActionView();
+        searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
+        searchView.setQuery(query, false);
+        android.widget.SearchView x;
+        searchView.setSubmitButtonEnabled(false);
+        searchView.setOnQueryTextListener(this);
+        SearchSuggestionsAdapter adapter = new SearchSuggestionsAdapter(this, null, 0);
+        //searchView.setSuggestionsAdapter(adapter);
+        return true;
+    }
+
+    private void showConnectionWarning() {
+        error.setVisibility(View.VISIBLE);
+        Button retry = error.findViewById(R.id.retry_button);
+        retry.setOnClickListener(view -> error.setVisibility(View.GONE));
+    }
+
+    private void chooseAdapter() {
```

Comment:
Choose adapter for recycler-view depending on entity.

Reply:
Good!

I still don't know what parts where inherited, so I see things that I think could be moved to viewModel (like `doSearch`)

Reply:
I have written this part. As a rule of thumb, you may keep in mind that the inherited code was 10 years old. Most of the changes are refactoring related and the rest changes were those made to search and login.